cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -1182,22 +1184,25 @@ class AsofJoinNode : public ExecNode {
 
       for (int i = 0; i < input_schema[j]->num_fields(); ++i) {
         const auto field = input_schema[j]->field(i);
+        bool as_output;
         if (i == on_field_ix) {
           ARROW_RETURN_NOT_OK(is_valid_on_field(field));
           // Only add on field from the left table
-          if (j == 0) {
-            fields.push_back(field);
-          }
+          as_output = (j == 0);
         } else if (std_has(by_field_ix, i)) {
           ARROW_RETURN_NOT_OK(is_valid_by_field(field));
           // Only add by field from the left table
-          if (j == 0) {
-            fields.push_back(field);
-          }
+          as_output = (j == 0);
         } else {
           ARROW_RETURN_NOT_OK(is_valid_data_field(field));
+          as_output = true;
+        }
+        if (as_output) {
           fields.push_back(field);
         }
+        if (field_output_indices) {
+          field_output_indices->push_back(as_output ? output_field_idx++ : i);
```

Comment:
Can you give an example of how the field output indices looks like, e.g. for input tables `time,k,v1` and `time,k,v2` how does `field_output_indices` look like?

Reply:
For an as-of-join on `t` by `k` with input tables `t,k,v1` and `t,k,v2`, the resulting `field_output_indices` will be `0, 1, 2, 0, 1, 3`. Of these, the 4th and 5th indices (with values 0 and 1) have `as_output` set to `false`, indicating a non-output field.

Your question made me realize there is a bug here. Instead of setting to `i`, which is an index of a non-output field in a right table, it should be setting to the corresponding index of the left table, since the on- and by-indices are output only for the left table, and in general their indices in the right and left tables may not be the same. I'll fix this.

Reply:
Fixed.